A51-custom-backend-metrics.md

```diff
@@ -380,21 +423,131 @@ public interface OrcaOobReportListener {
+
+  // Records a call metric measurement for queries per second.
+  // Multiple calls to this method will override the stored value.
+  // Values outside of the valid range [0, infy) are ignored.
+  CallMetricRecorder& RecordQpsMetric(double value);
+
+  // Records a call metric measurement for utilization.
+  // Multiple calls to this method with the same name will
+  // override the corresponding stored value. The lifetime of the
+  // name string needs to be longer than the lifetime of the RPC
+  // itself, since it's going to be sent as trailers after the RPC
+  // finishes. It is assumed the strings are common names that
+  // are global constants.
+  // Values outside of the valid range [0, 1] are ignored.
+  CallMetricRecorder& RecordUtilizationMetric(string_ref name,
+                                                      double value);
+
+  // Records a call metric measurement for request cost.
+  // Multiple calls to this method with the same name will
+  // override the corresponding stored value. The lifetime of the
+  // name string needs to be longer than the lifetime of the RPC
+  // itself, since it's going to be sent as trailers after the RPC
+  // finishes. It is assumed the strings are common names that
+  // are global constants.
+  CallMetricRecorder& RecordRequestCostMetric(string_ref name,
+                                                      double value);
+};
 
-TODO(nicolasnoble): Fill this in.
+class ServerContextBase {
+ public:
+  // Get the a CallMetricRecorder object for the current RPC.
+  // Use it to record metrics during your RPC to send back to the
+  // client in order to make load balancing decisions. This will
+  // return nullptr if the feature hasn't been enabled using
+  // EnableCallMetricRecording().
+  CallMetricRecorder* GetCallMetricRecorder();
+};
+```
 
-#### OOB Reporting APIs
+#### Per-Server Recording on Server
 
-C++ provides the following API for running the ORCA service:
+C++ will provide `ServerMetricRecorder` to record metrics per-server.
+The user creates an instance and records metrics to it. All the methods are thread safe.
+This is optional for per-request load reporting and required for OOB load reporting.
+```c++
+class ServerMetricRecorder {
+ public:
+  // Factory method. Use this to create.
+  static std::unique_ptr<ServerMetricRecorder> Create();
+  // Records the server CPU utilization in the range [0, 1].
+  // Values outside of the valid range are rejected.
+  // Overrides the stored value when called again with a valid value.
+  void SetCpuUtilization(double value);
+  // Records the server memory utilization in the range [0, 1].
+  // Values outside of the valid range are rejected.
+  // Overrides the stored value when called again with a valid value.
+  void SetMemoryUtilization(double value);
+  // Records number of queries per second to the server in the range [0, infy).
+  // Values outside of the valid range are rejected.
```

Comment:
Looking at the [WIP grpc java implementation](https://github.com/grpc/grpc-java/pull/9902), they are ignored, not rejected. Should we clarify here if by "rejected" we mean "ignored" - because rejected may imply the method could throw.

On the other hand, I'm a bit afraid if we just ignore the value. I agree that throwing is a bad idea, but just swallowing an error leads to reporting incorrect data. I don't think we can judge if skewing a particular metric is critical to a customer.

One example that come to mind: a customer is storing qps internally in an int, the service gets a huge number of requests, and the int gets overflowed. In this case, they'll call `setQps()` with a negative value, when in fact there was a huge spike. If another service goes down because of a qps spike, but our service doesn't report the qps increase, an engineer responding to the incident won't be able to identify the source of the request spike correctly.

I wonder if it'd make sense require logging a warning when an incorrect value is presented to the MetricRecorder/CallMetricRecorder. However, depending on how often the data is reported, it might lead to the logs getting flooded. Obviously, we can restrict how often to produce these warnings, but this may be going to far out-of-scope.

Reply:
This section is for C++ and it's from comments in code so I don't think "rejected" may confuse. Java snippet does not have this word.

Regarding the range validation, I believe the caller can validate or log in such cases. Note that recorder APIs already ignore with no feedback today (before this change) and this PR just clarifies that. Today the value will be rejected during proto's range validation and/or when consumed by the client where neither is explicit.